### PR TITLE
Reapply PR/87550

### PR DESCRIPTION
**Pull request by oontvoo**
Re-apply https://github.com/llvm/llvm-project/pull/87550 with fixes.

Details:
Some tests in fuchsia failed because of the newly added assertion.
This was because `GetExceptionBreakpoint()` could be called before `g_dap.debugger` was initted.

The fix here is to just lazily populate the list in GetExceptionBreakpoint() rather than assuming it's already been initted.
(There is some nuisance here because we can't simply just populate it in DAP::DAP(), which is a global ctor and is called before `SBDebugger::Initialize()` is called. )

### CODE REVIEW

File: lldb/tools/lldb-dap/DAP.cpp [hunk 1]
```diff
@@ -65,16 +58,60 @@ DAP::DAP()
...
+  // We also cannot call PopulateExceptionBreakpoints() in DAP::DAP() because
+  // we need SBDebugger::Initialize() to have been called before this.
+  //
+  // So just checking the filter list and do lazy-populating seems easiest.
+  // Two other options include:
+  //  + call g_dap.PopulateExceptionBreakpoints() in lldb-dap.cpp::main()
+  //    right after the call to SBDebugger::Initialize()
+  //  + Just call PopulateExceptionBreakpoints() to get a fresh list  everytime
+  //    we query (a bit overkill since it's not likely to change?)
+  if (!exception_breakpoints.has_value())
```
[the remark this answers is not in the document]
Comment (oontvoo) [reply lldb/tools/lldb-dap/DAP.cpp]: Thanks! I've removed the `if()` check and wrap the block of code that populates `exception_breakpoints` in a call_once so the callers of `PopulateExceptionBreakpoints` don't have to do that

File: lldb/tools/lldb-dap/DAP.cpp [hunk 2]
```diff
@@ -59,25 +59,29 @@ DAP::DAP()
...
-                                        lldb::eLanguageTypeObjC);
-  }
-  if (lldb::SBDebugger::SupportsLanguage(lldb::eLanguageTypeSwift)) {
-    exception_breakpoints->emplace_back("swift_catch", "Swift Catch",
-                                        lldb::eLanguageTypeSwift);
-    exception_breakpoints->emplace_back("swift_throw", "Swift Throw",
-                                        lldb::eLanguageTypeSwift);
-  }
+  if (exception_breakpoints.has_value())
+    return;
```
Comment (JDevlieghere) [inline lldb/tools/lldb-dap/DAP.cpp:63]: The `call_once` is already going to guarantee that this isn't called more than once, so this check is redundant.

**Review by JDevlieghere — APPROVED** [review]
LGTM without the redundant check.

File: lldb/tools/lldb-dap/DAP.cpp (hunk 2 above)
Comment (oontvoo) [reply lldb/tools/lldb-dap/DAP.cpp]: thanks!

**Comment by felipepiovezan** [comment]
Oops, I commented on the old PR instead of the new one, so let me copy paste it here:

@oontvoo I think this broke the bots again: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/5341/console
Would you mind reverting it until you have a chance to look at the failures?

**Comment by felipepiovezan** [comment]
@oontvoo I'm going to revert this for now as the incremental bots are our first line of defense against failures

**Pull request merged**